packages/actor-context-preprocess-rdf-source-identifier/lib/ActorContextPreprocessRdfSourceIdentifier.ts

```diff
@@ -56,6 +56,11 @@ export class ActorContextPreprocessRdfSourceIdentifier extends ActorContextPrepr
         }
       }
 
+      // If the sources are fixed, block until all sources are transformed.
+      if (sources.isEnded()) {
+        await new Promise((resolve) => it.on('end', resolve));
```

Comment:
It's not a fix for a bug :-)

It's just a way to make the `run` promise block until the `sources` iterable ends. But this will _only_ happen in the case that `sources` was fixed before (not infinite). This makes some things later down the pipeline easier to handle. If it was infinite, then this transformation will happen async.

Reply:
The thing I'm wondering is how is it possible that `sources.isEnded()` is true, but `it` has not thrown an 'end' event yet, since `it === sources.iterator()`

Reply:
That is because `it` is being handled asynchronously. In most cases, the on-data stuff will be processed behind a `setImmediate`. That means that our `run` will fully run, and only after that, the on-data and on-end callbacks will be called. The line above ensures that `run` 'blocks' until these on-data and on-end things are called.